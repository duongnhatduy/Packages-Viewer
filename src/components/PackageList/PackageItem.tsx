import React, { useState } from "react";
import { TPackage } from "./types";

type TPackageItemProps = {
  package: string;
  list: {
    [name: string]: TPackage;
  };
};

export default function PackageItem({
  package: name,
  list
}: TPackageItemProps) {
  const [isExpanded, setExpanded] = useState(false);
  const currentPackage = list[name];
  const { depends = "", description = "" } = currentPackage || {};

  if (!currentPackage) {
    return <div>{name} not installed</div>;
  }

  return (
    <>
      <button onClick={() => setExpanded(!isExpanded)}>{name}</button>
      {isExpanded && (
        <div>
          {description}
          {!!depends && (
            <>
              <h2>Dependencies</h2>
              <ul>
                {getDependencies(depends).map(dependency => (
                  <li key={dependency}>
                    <PackageItem package={dependency} list={list} />
                  </li>
                ))}
              </ul>
            </>
          )}
        </div>
      )}
    </>
  );
}

/**
 *
 * @param depends
 * @description get dependencies name from depends field
 * @example "libc6 (>= 2.7), upstart-job" will return ["libc6", "upstart-job"]
 */
function getDependencies(depends: string): string[] {
  return depends.split(", ").map(dependency => dependency.split(" ")[0]);
}
