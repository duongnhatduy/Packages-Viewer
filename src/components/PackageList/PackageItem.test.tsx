import React from "react";
import {
  render,
  fireEvent,
  cleanup,
  waitForElement
} from "@testing-library/react";
import "jest-dom/extend-expect";

import PackageItem from "./PackageItem";

afterEach(cleanup);

const sample = {
  acpid: {
    package: "acpid",
    depends:
      "libc6 (>= 2.7), upstart-job, lsb-base (>= 3.2-14), module-init-tools (>> 3.1-rel-2)",
    description:
      "Advanced Configuration and Power Interface event daemon\n Modern computers support the Advanced Configuration and Power Interface (ACPI)\n to allow intelligent power management on your system and to query battery and\n configuration status.\n .\n ACPID is a completely flexible, totally extensible daemon for delivering\n ACPI events. It listens on netlink interface (or on the deprecated file\n /proc/acpi/event), and when an event occurs, executes programs to handle the\n event. The programs it executes are configured through a set of configuration\n files, which can be dropped into place by packages or by the admin."
  },
  libc6: {
    package: "libc6",
    depends: "libc-bin (= 2.15-0ubuntu10.3), libgcc1, tzdata",
    description: "Embedded GNU C Library"
  }
};

function setup(testProps: Partial<React.ComponentProps<typeof PackageItem>>) {
  const wrapper = render(
    <PackageItem list={sample} package="" {...testProps} />
  );
  return { ...wrapper };
}

test("renders correct expanded dependency", () => {
  const { getByText, queryByText, container } = setup({ package: "acpid" });
  const btn = getByText("acpid");
  expect(btn).toBeVisible();

  expect(queryByText(/Advanced Configuration and Power Interface/)).toBeNull();

  // click to expand package
  fireEvent.click(btn);
  expect(getByText(/Advanced Configuration and Power Interface/)).toBeVisible();
  expect(getByText(/libc6/)).toBeVisible();
  expect(getByText(/lsb-base not installed/)).toBeVisible();

  fireEvent.click(getByText(/libc6/));
  expect(getByText(/Embedded GNU C Library/)).toBeVisible();
  expect(container).toMatchSnapshot();
});

test("renders with reverse dependencies", () => {
  const { getByText, container } = setup({ package: "libc6" });
  const btn = getByText("libc6");
  expect(btn).toBeVisible();
  fireEvent.click(btn);

  expect(getByText(/acpid/)).toBeVisible();
  fireEvent.click(getByText(/acpid/));
  expect(getByText(/Advanced Configuration and Power Interface/)).toBeVisible();
  expect(container).toMatchSnapshot();
});
