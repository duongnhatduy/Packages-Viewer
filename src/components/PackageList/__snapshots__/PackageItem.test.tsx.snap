// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correct dependency 1`] = `
<div>
  <button>
    acpid
  </button>
  <div>
    Advanced Configuration and Power Interface event daemon
 Modern computers support the Advanced Configuration and Power Interface (ACPI)
 to allow intelligent power management on your system and to query battery and
 configuration status.
 .
 ACPID is a completely flexible, totally extensible daemon for delivering
 ACPI events. It listens on netlink interface (or on the deprecated file
 /proc/acpi/event), and when an event occurs, executes programs to handle the
 event. The programs it executes are configured through a set of configuration
 files, which can be dropped into place by packages or by the admin.
    <section>
      <h2>
        Dependencies
      </h2>
      <ul>
        <li>
          <button>
            libc6
          </button>
          <div>
            Embedded GNU C Library
            <section>
              <h2>
                Dependencies
              </h2>
              <ul>
                <li>
                  <div>
                    libc-bin
                     not installed
                  </div>
                </li>
                <li>
                  <div>
                    libgcc1
                     not installed
                  </div>
                </li>
                <li>
                  <div>
                    tzdata
                     not installed
                  </div>
                </li>
              </ul>
            </section>
            <section>
              <h2>
                Reverse dependencies
              </h2>
              <ul>
                <li>
                  <button>
                    acpid
                  </button>
                </li>
              </ul>
            </section>
          </div>
        </li>
        <li>
          <div>
            lsb-base
             not installed
          </div>
        </li>
        <li>
          <div>
            module-init-tools
             not installed
          </div>
        </li>
        <li>
          <div>
            upstart-job
             not installed
          </div>
        </li>
      </ul>
    </section>
    <section>
      <h2>
        Reverse dependencies
      </h2>
      <ul />
    </section>
  </div>
</div>
`;

exports[`renders correct expanded dependency 1`] = `
<div>
  <button>
    acpid
  </button>
  <div>
    Advanced Configuration and Power Interface event daemon
 Modern computers support the Advanced Configuration and Power Interface (ACPI)
 to allow intelligent power management on your system and to query battery and
 configuration status.
 .
 ACPID is a completely flexible, totally extensible daemon for delivering
 ACPI events. It listens on netlink interface (or on the deprecated file
 /proc/acpi/event), and when an event occurs, executes programs to handle the
 event. The programs it executes are configured through a set of configuration
 files, which can be dropped into place by packages or by the admin.
    <section>
      <h2>
        Dependencies
      </h2>
      <ul>
        <li>
          <button>
            libc6
          </button>
          <div>
            Embedded GNU C Library
            <section>
              <h2>
                Dependencies
              </h2>
              <ul>
                <li>
                  <div>
                    libc-bin
                     not installed
                  </div>
                </li>
                <li>
                  <div>
                    libgcc1
                     not installed
                  </div>
                </li>
                <li>
                  <div>
                    tzdata
                     not installed
                  </div>
                </li>
              </ul>
            </section>
            <section>
              <h2>
                Reverse dependencies
              </h2>
              <ul>
                <li>
                  <button>
                    acpid
                  </button>
                </li>
              </ul>
            </section>
          </div>
        </li>
        <li>
          <div>
            lsb-base
             not installed
          </div>
        </li>
        <li>
          <div>
            module-init-tools
             not installed
          </div>
        </li>
        <li>
          <div>
            upstart-job
             not installed
          </div>
        </li>
      </ul>
    </section>
    <section>
      <h2>
        Reverse dependencies
      </h2>
      <ul />
    </section>
  </div>
</div>
`;

exports[`renders with reverse dependencies 1`] = `
<div>
  <button>
    libc6
  </button>
  <div>
    Embedded GNU C Library
    <section>
      <h2>
        Dependencies
      </h2>
      <ul>
        <li>
          <div>
            libc-bin
             not installed
          </div>
        </li>
        <li>
          <div>
            libgcc1
             not installed
          </div>
        </li>
        <li>
          <div>
            tzdata
             not installed
          </div>
        </li>
      </ul>
    </section>
    <section>
      <h2>
        Reverse dependencies
      </h2>
      <ul>
        <li>
          <button>
            acpid
          </button>
          <div>
            Advanced Configuration and Power Interface event daemon
 Modern computers support the Advanced Configuration and Power Interface (ACPI)
 to allow intelligent power management on your system and to query battery and
 configuration status.
 .
 ACPID is a completely flexible, totally extensible daemon for delivering
 ACPI events. It listens on netlink interface (or on the deprecated file
 /proc/acpi/event), and when an event occurs, executes programs to handle the
 event. The programs it executes are configured through a set of configuration
 files, which can be dropped into place by packages or by the admin.
            <section>
              <h2>
                Dependencies
              </h2>
              <ul>
                <li>
                  <button>
                    libc6
                  </button>
                </li>
                <li>
                  <div>
                    lsb-base
                     not installed
                  </div>
                </li>
                <li>
                  <div>
                    module-init-tools
                     not installed
                  </div>
                </li>
                <li>
                  <div>
                    upstart-job
                     not installed
                  </div>
                </li>
              </ul>
            </section>
            <section>
              <h2>
                Reverse dependencies
              </h2>
              <ul />
            </section>
          </div>
        </li>
      </ul>
    </section>
  </div>
</div>
`;
